### DISCUSSION TITLE
How to properly send validation errors and how to access it in the js ClientResponseError?

### DISCUSSION THREAD

**Discussion by mukhtharcm** [pocketbase/pocketbase#5535] [Q&A]
I tried sending 
```go
// print the error
fmt.Println(err.Error())
return apis.NewBadRequestError(err.Error(), nil)
```

but the data field on clientresponse error is empty :)

**Comment by ganigeorgiev** [comment]
Please check the example in https://pocketbase.io/docs/go-routing/#error-response.

Only `validation.Errors` or `map[string]validation.Error` are considered public safe.

**Comment by mukhtharcm** [comment]
I changed according to this.

```go
app.OnModelBeforeCreate("program_candidates").Add(func(e *core.ModelEvent) error {
		err := checkProgramLimits(app, e.Model.(*models.Record))
		if err != nil {
			// print the error
			fmt.Println(err.Error())
			return apis.NewApiError(http.StatusBadRequest, err.Error(), map[string]validation.Error{
				"program": validation.NewError("max_limit", err.Error()),
			})
		}
		return nil
	})
```

th error is printing properly, but still empty data on clientreponseerror.
Do you know what I'm doing wrong?

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean. The above should work correctly. Make sure that you are inspecting the correct error prop (e.g. `console.log(err.response)`).

Also note that that the above error is equivalent to:

```go
apis.NewBadRequestError(err.Error(), map[string]validation.Error{
    "program": validation.NewError("max_limit", err.Error()),
})
```

If you need further help, please try to elaborate more clearly what actually is the problem and how do you check for the error in the first place to avoid the pointless back-and-forth.

**Reply by mukhtharcm** [reply]
<img width="696" alt="image" src="https://github.com/user-attachments/assets/e2da7bd0-41a5-4046-a426-a26f9e3fbafd">
it's also showing properly In the Logs

**Reply by ganigeorgiev** [reply]
Oh, I just noticed that you are trying to return an `ApiError` from a model DB hook. This is not supported at the moment and `ApiError` could be returned only from `*Request` hooks or middlewares/route handlers (_support for unwrapping model hook errors is already implemented with the ongoing refactoring and will be available in the next release_).

**Reply by ganigeorgiev** [reply]
See also https://github.com/pocketbase/pocketbase/discussions/3464.

**Reply by mukhtharcm** [reply]
Thank you for pointing that out!
So currently there's no way for surfacing that to UI.
I think It'll be good to make a custom route for doing this for now :)

Thanks for doing this good work 💌

**Reply by ganigeorgiev** [reply]
Yes, ApiError cannot be returned from non-Request related hooks at the moment. If you just want to return the validation error, then you could try returning from the hook `validation.Errors` but it still is not guarantee that it will work if the error is wrapped by some other hook:

```go
return validation.Errors{
    "program": validation.NewError("max_limit", err.Error()),
}
```

**Reply by Odas0R** [reply]
> Yes, ApiError cannot be returned from non-Request related hooks at the moment. If you just want to return the validation error, then you could try returning from the hook `validation.Errors` but it still is not guarantee that it will work if the error is wrapped by some other hook:
> 
> ```go
> return validation.Errors{
>     "program": validation.NewError("max_limit", err.Error()),
> }
> ```

This works, thanks :)
